juju/juju
# [JUJU 622] Fix origin for container addresses

As part of the upgrade steps for 2.8, we added `AddOriginToIPAddresses`. This was in order to accommodate changed behaviour to coordinate between address update logic in instance-poller and the machine agent.

Addresses have an _origin_ to determine who is responsible for an address.
- If the instance-poller does not see an address, it does not remove it. It instead sets the origin to `machine`, in effect relinquishing control of the address.
- If the machine agent does not see an address, it will only remove it if the origin is `machine`, so it does not remove addresses only known by the instance poller (such as public addresses).

The upgrade step set all address origins to `provider` under the assumption that they would go through their natural attrition process, and be removed if no longer against a particular device. However, we missed that the instance-poller does not run for container addresses, and so we observed that IPs assigned by Corosync/Pacemaker were "frozen" by upgrading, where such addresses were against containers - they could never be relinquished to `machine` control for deletion.

Here we add an upgrade step to update the origin of LXD containers and KVMs to ensure that they are under `machine` control, allowing them to be removed if no longer detected.

## QA steps

- Bootstrap current 2.9 to a substrate allowing machine-in-machine deployment (I used ServerStack).
- `juju add machine`
- `juju add machine lxd:0`
- `juju ssh 0`, then `lxc exec <name> bash` to get into the container.
- Add a secondary address to the main NIC, for example `ip address add 252.3.234.99/8 dev eth0`.
- Then restart the machine agent to expedite detection: `systemctl restart jujud-machine-0-lxd-0.service`.
- Now change the origin of the newly added address to be "provider". Example: `db.ip.addresses.updateOne({"_id" : "967b5cc4-8b6c-4eba-86d9-1a62979dd150:m#0/lxd/0#d#eth0#ip#252.3.234.99"}, {$set: {"origin": "provider"}})`.
- Access the container as above and delete the address: `ip address del 252.3.234.99/8 dev eth0`.
- (Optional) if you restart the machine agent again, you will note that the address is not deleted.
- Now apply the patch below to this one, to ensure 2.9.26 upgrade steps will run:
```diff
diff --git a/version/version.go b/version/version.go
index ebefe321ce..3913bc548c 100644
--- a/version/version.go
+++ b/version/version.go
@@ -21,7 +21,7 @@ import (
 // The presence and format of this constant is very important.
 // The debian/rules build recipe uses this value for the version
 // number of the release package.
-const version = "2.9.25"
+const version = "2.9.26"
 
 // UserAgentVersion defines a user agent version used for communication for
 // outside resources.
```
- `juju upgrade-controller --build-agent`, then `juju upgrade-model`.
- Observe that the address you added has been removed.

## Documentation changes

None.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1944758
